TelemetryDesignDoc.md

```diff
@@ -0,0 +1,327 @@
+
+### Verbose Shuffleboard API
+
+The Shuffleboard API is extremely powerful and allows more or less the entire dashboard layout to be declaratively structured from robot code.  The native API for this, however, is verbose and not self-documenting.  Simple declarations require long method chains, and widget arguments are passed via string key-value pairs without type safety.
+
+### Lack of a universal metadata layer
+
+Both the new datalogging implementation and NT4 include support for string-valued metadata.  The implementation of the code-driven layout configuration in Shuffleboard involves an ad-hoc formatting metadata layer that is not shared by any of the other tooling.
+
+Since these metadata layers are incompatible, any effort spent on Shuffleboard widgets is useless for the newer Glass-based tooling.  Were these metadata layers compatible (ideally, identical), then Shuffleboard features can smoothly interop with Glass and other more-modern dashboards (including, potentially, a web-based successor to Shuffleboard).
+
+This seems to have been the original intent of Shuffleboard's metadata structure and adherence to generic string-string maps for widget properties.  Building along these lines preserves much of the design work that went into the tool.
+
+### Obscure calls for simple functions
+
+The simplest way to log telemetry to the dashboard declaratively is through a `SmartDashboard` method, despite `SmartDashboard` being a nearly-obsolete legacy framework with no active maintenance.  As Glass does not have its own API, ahe `Shuffleboard` API is rather verbose, `SmartDashboard` calls are often used even in codebases that never actually open `SmartDashboard`.  This is not ideal.
+
+Moreover, the various `put` functions for both `SmartDashboard` and `Shuffleboard` suddenly change semantics from immediate-mode periodic "put" calls to declarative databinding depending on whether the argument is a `Sendable`.  This overloading of the `put` name is highly confusing.
+
+### Redundant APIs for defining subsystem/device properties
+
+A subsystem (or device--there is very little difference at this level) currently needs to declare its properties multiple times: once for telemetry/NT, once for simulation, and possibly more. NT and simulation being relatively simialar isn't new, and is evident from classes such as `Field2d` and `Mechanism2d` being migrated rather easily from a simulation backend to an NT one. The WS setup used primarily for the Romi bots is also very similar to NT and sim.
+
+While a longer-term goal, merging of NT and simulation would decrease boilerplate in both team and library code. Developing sim extensions/plugins would also become easier, partially due to HALSim extensions currently needing to be written in C++ while a connection to NT can be implemented in nearly any programming language (for starters, NT APIs exist for Java, Python, JS, and more).
```

Comment:
I don't understand what you're getting at re: duplicated structure.  In order to simulate a robot, somewhere the user has to describe the physical behavior of the robot mechanisms (e.g. what motors change what sensors, the motor type and output/sensor gear ratios, the device/world physics so the feedback sensors read realistic values, etc).  None of these things are necessary to specify to just drive a motor with a joystick, and I don't think we want to *require* that they be.

Reply:
It's referring to the HALSim, not the physics sim. Classes representing devices declare properties to the NT API (for telemetry) as well as the SimDevice/HALSim API. The catch is that the two sets of properties are identical in most cases, and therefore I think the duplicate declaration of properties can be reworked to have a single declaration.

Reply:
The HAL simulation layer is designed to be synchronous with user code to enable exactly reproducible sim behavior--it provides sim callbacks that are called in line with the user code.  NT will never have this behavior.  It's also automatically stubbed out on a real robot in a way that NT isn't.

Reply:
Was there desire to align with the ROS paradigm? IE, if you made your robot code and simulation and hardware communication into multiple, independent but communicating NT clients, you could in theory swap out real hardware nodes with simulation?

I guess I see that as a far-out architecture and, for the purposes of 75% of FRC, not extremely useful. Still nifty to consider though, and I don't see anything in the doc that boxes a team out from doing it.

Reply:
I think you're misinterpreting this as merging of implementation, rather than merging of declaration.  The point is to reduce  DRY issues caused by double-declaring large parts of the robot structure.  I'm updating the prose to clarify this.